### ISSUE TITLE
vBox is inconsistent

### ISSUE THREAD

**Issue by siddhanathan** [jtdaugherty/brick#133]
I'm using `brick-0.18` and `vty-5.16`.

Here's a small snippet of code I have:

```haskell
boardNumbers :: Game -> [[Widget Name]]

drawBoard :: Game -> Widget Name
drawBoard g = (border . vBox . fmap hBox) (boardNumbers g)
```

Which produces a UI similar to the following:

![screenshot from 2017-08-21 03-38-29](https://user-images.githubusercontent.com/933964/29509576-b90dddca-8626-11e7-9e98-d28fc4ef4f8b.png)

The first row seems to be a bit bigger than the rest. There are certain screen sizes where the first row height is much more consistent, but most of the time the first row seems to be slightly bigger than the rest. I'm guessing that this is an issue in the library, and not in my code.

**Comment by jtdaugherty** [comment]
Thanks for the report! Would it be possible for me to see the whole program? Or could you create a minimal demonstration program that has this behavior? `vBox`'s behavior depends greatly on the things in the box and how they use space, so `vBox` alone is not enough to trigger this problem.

**Comment by jtdaugherty** [comment]
Also, in case you are interested, I am working on an ASCII art editor program using `brick` that you could use to make some ASCII art checkers and then load them into your `brick` UI. :)

https://github.com/jtdaugherty/tart

**Comment by siddhanathan** [comment]
Thanks! I will definitely look into `tart`. I don't mind sharing the whole program if needed, but there's tightly coupled network code which will probably make debugging harder. Attaching a minimal example that has this behavior.

```haskell
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ViewPatterns #-}

module Main where

import Control.Monad (void)
import Data.Function ((&))
import Data.Function (on)
import Data.List     (transpose)

import qualified Brick                as B
import qualified Brick.Widgets.Border as B
import qualified Brick.Widgets.Center as B
import qualified Graphics.Vty         as V

type Name = ()
data Tick
data Game

app :: B.App Game Tick Name
app = B.App
    { B.appDraw         = drawUI
    , B.appChooseCursor = B.neverShowCursor
    , B.appHandleEvent  = handleEvent
    , B.appStartEvent   = pure
    , B.appAttrMap      = const theMap
    }

theMap :: B.AttrMap
theMap = B.attrMap V.defAttr
    [ (blackSquare, go V.withBackColor V.black)
    , (redSquare, go V.withBackColor V.red)
    ]
  where
    go a b = a V.defAttr b

blackSquare, redSquare :: B.AttrName
blackSquare = "blackSquare"
redSquare = "redSquare"

vkey :: B.BrickEvent Name Tick -> Maybe V.Key
vkey (B.VtyEvent (V.EvKey y [])) = Just y
vkey _ = Nothing

handleEvent :: Game -> B.BrickEvent Name Tick -> B.EventM Name (B.Next Game)
handleEvent g (vkey -> Just (V.KChar 'q')) = g & B.halt
handleEvent g (vkey -> Just (V.KEsc))      = g & B.halt
handleEvent g (vkey -> Nothing)            = g & B.continue
handleEvent g _                            = g & B.continue

drawUI :: Game -> [B.Widget Name]
drawUI = pure . drawBoard

boardNumbers :: Game -> [[B.Widget Name]]
boardNumbers _ =
    board
  where
    (.:) = (.) . (.)

    ap = take 8 .: zipC `on` repeat

    board    = ap oddRows evenRows
    oddRows  = ap reds    blacks
    evenRows = ap blacks  reds

    zipC = (concat . transpose) .: mappend `on` pure

    pattern p = B.withAttr p . B.center . B.str $ " "

    reds   = pattern redSquare
    blacks = pattern blackSquare

drawBoard :: Game -> B.Widget Name
drawBoard = B.border . B.vBox . fmap B.hBox . boardNumbers

main :: IO ()
main = void $ B.defaultMain app undefined
```

**Comment by jtdaugherty** [comment]
Thanks - that made it easy to see the behavior. Here's why you're seeing what you're seeing:

When you put things in a box and they're all greedy (in this case they are due to `center`), then the box attempts to divide the available space evenly between the elements. However, if there isn't enough space to divide it perfectly evenly, any leftover space is assigned to the first element. This is why you see a first row being higher than the others... for some terminal window sizes. If the window size is a perfect multiple of the number of cells in each column and row, they'll all have the same height and width.

Now, this is a fine behavior in cases where the "leftover" space is just one or two rows or columns, but I agree that in your case and any case like it, it's not a great behavior and it isn't obvious that it will happen. There are two alternatives:

* Whenever there is leftover room, spread it out evenly among as many elements as possible. So if there are two extra rows, assign one row each to the first and second box elements. This approach has the same drawback as the current implementation, but it's just less pronounced: some cells will be higher or wider than the rest, and if you resize the terminal window then the UI will lay out differently.

* Whenever there is leftover room, just don't use it. In other words, the UI will lay out so that all of the rows and columns have identical dimensions. This is nice because it avoids the issue entirely, but it is bad because it breaks the "greedy" contract that the box elements have, which says that they'll use the space available to them. It will result in UIs that are smaller than expected and will have extreme resize jitter on terminal window size changes.

I'm going to implement the first approach, which is a compromise between the second and the current implementation. If that doesn't work for you, then you can get around this entirely by using `hLimit` and `vLimit` to set the size of your board cells to your own dimensions. Additionally you can write a custom widget that gets the rendering context size and divides the available space up exactly as you see fit.

**Comment by jtdaugherty** [comment]
The first approach is now implemented in https://github.com/jtdaugherty/brick/commit/adf599ff4981325bdf4326cf258b754023f0784c.

**Comment by siddhanathan** [comment]
Wow, that was quick. Thanks! Works for me. For future reference, attaching a before and after screenshot.

![screenshot from 2017-08-22 02-51-38](https://user-images.githubusercontent.com/933964/29552501-ac93c19e-86e5-11e7-9692-4904157bfee2.png)

**Comment by jtdaugherty** [comment]
This fix is now released in `0.24.1`.
